### PR TITLE
[ECO-5009][CHA-RL1] Room ATTACH 

### LINKED ISSUE #21: [CHA-RL1] Implement Room ATTACH operation
- Implement all spec items that come under CHA-RL1.
- Implement related integration tests.
- Split from #6

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5009) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by sacOO7**
Fixed #21

### CODE REVIEW

File: chat-android/src/main/java/com/ably/chat/ErrorCodes.kt [hunk 1]
```diff
@@ -3,105 +3,106 @@ package com.ably.chat
 /**
  * Error codes for the Chat SDK.
  */
-object ErrorCodes {
```
Comment (ttypic) [inline chat-android/src/main/java/com/ably/chat/ErrorCodes.kt:6]: Why do you think enum here is better?

File: chat-android/src/main/java/com/ably/chat/RoomStatus.kt [hunk 2]
```diff
@@ -118,3 +162,60 @@ data class RoomStatusChange(
      */
     val error: ErrorInfo? = null,
 )
+
+open class RoomStatusEvenEmitter : EventEmitter<RoomLifecycle, RoomStatus.Listener>() {
```
Comment (ttypic) [inline chat-android/src/main/java/com/ably/chat/RoomStatus.kt:166]: here is typo in the name, should be `...EventEmitter`

File: chat-android/src/main/java/com/ably/chat/RoomStatus.kt [hunk 3]
```diff
@@ -118,3 +162,60 @@ data class RoomStatusChange(
...
+    override fun apply(listener: RoomStatus.Listener?, event: RoomLifecycle?, vararg args: Any?) {
+        try {
+            listener?.roomStatusChanged(args[0] as RoomStatusChange)
+        } catch (t: Throwable) {
+            Log.e("RoomEventEmitter", "Unexpected exception calling Room Status Listener", t)
+        }
+    }
+}
+
+class DefaultStatus(private val logger: LogHandler?) : InternalRoomStatus, RoomStatusEvenEmitter() {
```
Comment (ttypic) [inline chat-android/src/main/java/com/ably/chat/RoomStatus.kt:177]: It's not Kotlin idiomatic to use inheritance here. It's much better to use composition

File: chat-android/src/main/java/com/ably/chat/ErrorCodes.kt (hunk 1 above)
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/ErrorCodes.kt:6]: Having type for given `ErrorCode` enforces compile time check with error code reusability. 
Also, IDE doesn't give right suggestions when matching error codes when it's generic `Int`.
This makes it difficult to write correct assertions in code as well as tests.

Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/ErrorCodes.kt:6]: Also, in chat-js, [`ErrorCode` is represented as an enum](https://github.com/ably/ably-chat-js/blob/fccf81494b998d8977c88c9cac09dac9b3eb8c16/src/core/errors.ts#L6), this makes it super easy to correlate between two codebases.

File: chat-android/src/main/java/com/ably/chat/RoomStatus.kt (hunk 2 above)
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/RoomStatus.kt]: Fixed https://github.com/ably-labs/ably-chat-kotlin/pull/27/commits/ce9c0baf4085c8199771f89c29b3d60fe399db26

File: chat-android/src/main/java/com/ably/chat/RoomStatus.kt (hunk 3 above)
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/RoomStatus.kt]: Fixed https://github.com/ably-labs/ably-chat-kotlin/pull/27/commits/ce9c0baf4085c8199771f89c29b3d60fe399db26

File: chat-android/src/main/java/com/ably/chat/ErrorCodes.kt (hunk 1 above)
Comment (ttypic) [reply chat-android/src/main/java/com/ably/chat/ErrorCodes.kt:6]: Not sure I understand, could you give an example? Typescript's enums are nothing like Kotlin's enums. Error codes usually come from backend, and they come as `Int`. I currently don't understand what benefits we get, except we have to additionally serialize/deserialize `Int` into enum

**Comment by sacOO7** [comment]
Superseded by https://github.com/ably-labs/ably-chat-kotlin/pull/33

File: chat-android/src/main/java/com/ably/chat/ErrorCodes.kt (hunk 1 above)
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/ErrorCodes.kt:6]: Sure, I am closing this PR in favor of https://github.com/ably-labs/ably-chat-kotlin/pull/33. We can continue discussion for the same there 👍

**Pull request closed without merging**